File: src/core/observable/Disposer.ts
import { Disposable } from "./Disposable";
import { LogManager } from "../Logger";
import { array_remove } from "../util";

const logger = LogManager.get("core/observable/Disposer");

/**
 * Container for disposables.
 */
export class Disposer implements Disposable {
    private _disposed = false;
    private readonly disposables: Disposable[];

    /**
     * The amount of disposables contained in this disposer.
     */
    get length(): number {
        return this.disposables.length;
    }

    get disposed(): boolean {
        return this._disposed;
    }

    constructor(...disposables: Disposable[]) {
        this.disposables = disposables;
    }

    /**
     * Add a single disposable and return the given disposable.
     */
    add<T extends Disposable>(disposable: T): T {
        if (this.disposed) {
            disposable.dispose();
        } else {
            this.disposables.push(disposable);
        }

        return disposable;
    }

    /**
     * Insert a single disposable at the given index and return the given disposable.
     */
    insert<T extends Disposable>(index: number, disposable: T): T {
        if (this._disposed) {
            disposable.dispose();
        } else {
            this.disposables.splice(index, 0, disposable);
        }

        return disposable;
    }

    /**
     * Add 0 or more disposables.
     */
    add_all(...disposables: Disposable[]): this {
        if (this._disposed) {
            for (const disposable of disposables) {
                disposable.dispose();
            }
        } else {
            this.disposables.push(...disposables);
        }

        return this;
    }

    /**
     * Removes and disposes the given disposable.
     */
    remove(disposable: Disposable): void {
        array_remove(this.disposables, disposable);
        disposable.dispose();
    }

    /**
     * Disposes all held disposables.
     */
    dispose_all(): void {
        this.dispose_at(0, this.disposables.length);
    }

    /**
     * Disposes all held disposables.
     */
    dispose(): void {
        this.dispose_all();
        this._disposed = true;
    }

    dispose_at(index: number, amount = 1): void {
        for (const disposable of this.disposables.splice(index, amount)) {
            try {
                disposable.dispose();
            } catch (e) {
                logger.warn("Error while disposing.", e);
            }
        }
    }
}

File: core/src/commonMain/kotlin/world/phantasmal/core/disposable/Disposer.kt
package world.phantasmal.core.disposable

/**
 * Container for disposables that disposes the contained disposables when it is disposed.
 */
class Disposer(vararg disposables: Disposable) : TrackedDisposable() {
    private val disposables = mutableListOf(*disposables)

    /**
     * The amount of held disposables.
     */
    val size: Int get() = disposables.size

    fun <T : Disposable> add(disposable: T): T {
        require(!disposed) { "Disposer already disposed." }

        disposables.add(disposable)
        return disposable
    }

    /**
     * Add 0 or more disposables.
     */
    fun addAll(disposables: Iterable<Disposable>) {
        require(!disposed) { "Disposer already disposed." }

        this.disposables.addAll(disposables)
    }

    /**
     * Add 0 or more disposables.
     */
    fun addAll(vararg disposables: Disposable) {
        require(!disposed) { "Disposer already disposed." }

        this.disposables.addAll(disposables)
    }

    fun isEmpty(): Boolean = disposables.isEmpty()

    /**
     * Removes and disposes the given [disposable].
     */
    fun remove(disposable: Disposable) {
        disposables.remove(disposable)
        disposable.dispose()
    }

    /**
     * Removes and disposes [amount] disposables at the given [index].
     */
    fun removeAt(index: Int, amount: Int = 1) {
        repeat(amount) {
            disposables.removeAt(index).dispose()
        }
    }

    /**
     * Disposes all held disposables.
     */
    fun disposeAll() {
        disposables.forEach { it.dispose() }
        disposables.clear()
    }

    override fun internalDispose() {
        disposeAll()
        super.internalDispose()
    }
}
